pkg/sql/codegen/tensorflow/codegen.go

```diff
@@ -169,14 +169,81 @@ func IsKerasModel(estimator string) (bool, string) {
 	return false, fmt.Sprintf("tf.estimator.%s", estimator)
 }
 
-func validateAttributes(trainStmt *ir.TrainStmt) error {
+// TODO(shendiaomo): Make the optimizer related code more general and exported in `attribute.go` if other frameworks
+// than TensorFlow have to support python objects as model attributes.
+
+func attrIsOptimizer(attrKey string) bool {
+	switch attrKey {
+	case "model.optimizer", "model.dnn_optimizer", "model.linear_optimizer":
+		return true
+	}
+	return false
+}
+
+func setDefaultOptimizer(trainStmt *ir.TrainStmt, optimizerParamName string) {
+	// TODO(shendiaomo): Try to get the default value from the python `inspect` module instead of hard coding
+	defaultValue := "Adagrad" // Defaults to DNN with a single optimizer parameter
+	switch trainStmt.Estimator {
+	case "LinearClassifier", "LinearRegressor":
+		defaultValue = "Ftrl"
+	case "DNNLinearCombinedClassifier", "DNNLinearCombinedRegressor":
+		if optimizerParamName == "model.linear_optimizer" {
+			defaultValue = "Ftrl"
+		}
+	}
+	trainStmt.Attributes[optimizerParamName] = defaultValue
+}
+
+func constructOptimizers(trainStmt *ir.TrainStmt) {
+	optimizerArgs := map[string]map[string]interface{}{}
+	for k, v := range trainStmt.Attributes {
+		if attrIsOptimizer(k) {
+			if optimizerArgs[k] == nil {
+				optimizerArgs[k] = map[string]interface{}{}
+			}
+
+		}
+		pieces := strings.Split(k, ".")
+		if len(pieces) == 3 {
```

Comment:
else return some optimizer format error?

Reply:
That's not necessary. If the attribute is not an optimizer, it will stay untouched thus raise an error in `initializeAttributes`.